R/utilities.R

```diff
@@ -65,15 +65,68 @@
     }
 }
 
-.splitX <- function(X, workers, tasks) 
+.split_X_redo <- function(X, redo_index, works_per_task, n){
+    # browser()
+    prealloc <- missing(n)
+    redo_index <- c(redo_index, !tail(redo_index, 1))
+    splittedX <- vector("list", ifelse(prealloc, 0L, n))
+    skip_num <- NA
+    last_redo_status <- redo_index[[1]]
+    task_i <- 1L
+    x_start <- 1L
+    x_len <- 0L
+    for(i in seq_along(redo_index)){
+        redo <- redo_index[[i]]
+        is_switch <- xor(last_redo_status, redo)
+        is_switch <- ifelse(is_switch || !redo, is_switch, x_len >= works_per_task)
+        if(!prealloc && is_switch){
+            if(last_redo_status){
+                splittedX[[task_i]] <- X[seq.int(x_start, length.out = x_len)]
+            }else{
+                splittedX[[task_i]] <- .rng_bploop_iter(x_len)
+            }
+        }
+        ## do the switch
+        if(is_switch){
+            last_redo_status <- redo
+            task_i <- task_i + 1L
+            x_start <- ifelse(redo, x_start, x_start + x_len)
+            x_len <- 1L
+        }else{
+            x_len <- x_len + 1
+        }
+        if(!prealloc && task_i > n)
+            break
+    }
+    if(prealloc){
+        task_i - tail(redo_index, 1) - 1L
+    }else{
+        splittedX
+    }
+}
+
+.splitX <- function(X, workers, tasks, redo_index = NULL)
 {
     if (tasks == 0L) {
         tasks <- workers
     } else {
         tasks <- min(length(X), tasks)
     }
-    idx <- .splitIndices(length(X), tasks)
-    relist(X, idx)
+    ## If redo index presents, split X based on the index while
+    ## preserving the correct rng stream.
+    ## it will respect `tasks` value
+    if (length(redo_index)){
+        ## Two-pass algorithm, the first pass calculate
+        ## the required list size to allocate, then do
+        ## the real allocation
+        works_per_task <- ceiling(length(X)/max(tasks, 1L))
```

Comment:
Are these `elements_per_task`?

Reply:
Yes, I call each element in the task a "work", but we can use "element" as well. I will rename it.